public/app/features/dashboard-scene/serialization/layoutSerializers/utils.ts

```diff
@@ -219,10 +215,25 @@ export function getRuntimeVariableDataSource(variable: QueryVariableKind): DataS
   return datasource;
 }
 
+export function getRuntimePanelDataSource(query: PanelQueryKind): DataSourceRef | undefined {
+  let datasource: DataSourceRef | undefined = undefined;
+  // if we don't have a datasource set, we will try to infer it based on the query kind
+  if (!query.spec.datasource?.uid) {
+    const defaultDatasource = config.bootData.settings.defaultDatasource;
+    const dsList = config.bootData.settings.datasources;
+    // this is look up by type
+    const bestGuess = Object.values(dsList).find((ds) => ds.meta.id === query.spec.query.kind);
+    datasource = bestGuess ? { uid: bestGuess.uid, type: bestGuess.meta.id } : dsList[defaultDatasource];
+  } else {
+    datasource = query.spec.datasource;
+  }
+  return datasource;
+}
```

Comment:
A similar code is used in two more place. can we unify?

Reply:
@dprokop I did the refactor here [`c127dad` (#103885)](https://github.com/grafana/grafana/pull/103885/commits/c127dad0845fc2ccc394ba80a33456637ad14633)